File: tests/agent_workloads/adapters/http_api.py
import json
import sys
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from threading import Thread
from urllib.error import HTTPError
from urllib.request import Request, urlopen


if len(sys.argv) != 2:
    raise SystemExit("usage: http_api INPUT_FILE")


class Handler(BaseHTTPRequestHandler):
    def do_POST(self):
        length = int(self.headers.get("Content-Length", "0"))
        try:
            payload = json.loads(self.rfile.read(length))
        except (TypeError, ValueError, json.JSONDecodeError):
            status, body = 400, "invalid-json"
        else:
            if self.path != "/agent":
                status, body = 404, "not-found"
            elif payload.get("name") == "":
                status, body = 422, "invalid-name"
            else:
                status, body = 200, f"accepted|{payload['name']}|{payload['role']}"
        encoded = body.encode()
        self.send_response(status)
        self.send_header("Content-Length", str(len(encoded)))
        self.end_headers()
        self.wfile.write(encoded)

    def log_message(self, _format, *_args):
        pass


server = ThreadingHTTPServer(("127.0.0.1", 0), Handler)
thread = Thread(target=server.handle_request)
thread.start()
try:
    request = Request(
        f"http://127.0.0.1:{server.server_port}/agent",
        data=Path(sys.argv[1]).read_bytes(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urlopen(request) as response:
            status, body = response.status, response.read().decode()
    except HTTPError as error:
        status, body = error.code, error.read().decode()
    print(f"{status}|{body}")
finally:
    thread.join(timeout=2)
    server.server_close()

File: tests/agent_workloads/adapters/http_api.mjs
import { readFileSync } from "node:fs";
import { createServer, request as httpRequest } from "node:http";

if (process.argv.length !== 3) {
  throw new Error("usage: http_api INPUT_FILE");
}

function responseFor(body) {
  try {
    const payload = JSON.parse(body);
    if (payload.name === "") return [422, "invalid-name"];
    return [200, `accepted|${payload.name}|${payload.role}`];
  } catch {
    return [400, "invalid-json"];
  }
}

const server = createServer((request, response) => {
  const chunks = [];
  request.on("data", (chunk) => chunks.push(chunk));
  request.on("end", () => {
    const [status, body] = request.url === "/agent" && request.method === "POST"
      ? responseFor(Buffer.concat(chunks).toString("utf8"))
      : [404, "not-found"];
    response.statusCode = status;
    response.setHeader("Content-Length", Buffer.byteLength(body));
    response.end(body);
  });
});

await new Promise((resolve) => server.listen(0, "127.0.0.1", resolve));
try {
  const body = readFileSync(process.argv[2]);
  const result = await new Promise((resolve, reject) => {
    const request = httpRequest(
      {
        hostname: "127.0.0.1",
        port: server.address().port,
        path: "/agent",
        method: "POST",
        headers: { "Content-Type": "application/json", "Content-Length": body.length },
      },
      (response) => {
        const chunks = [];
        response.on("data", (chunk) => chunks.push(chunk));
        response.on("end", () => resolve([response.statusCode, Buffer.concat(chunks).toString("utf8")]));
      },
    );
    request.on("error", reject);
    request.end(body);
  });
  console.log(`${result[0]}|${result[1]}`);
} finally {
  await new Promise((resolve) => server.close(resolve));
}
